mindmeld/active_learning/alp.py

```diff
@@ -0,0 +1,234 @@
+            self.data_bucket = deepcopy(self.init_data_bucket)
+            num_iterations = math.ceil(
+                len(self.data_bucket.unsampled_queries) / self.batch_size
+            )
+            iteration = 0
+            while iteration <= num_iterations:
+                logger.info(
+                    "Strategy: %s. Epoch: %s. Iteration %s.", strategy, epoch, iteration
+                )
+                logger.info(
+                    "Sampled Queries: %s", len(self.data_bucket.sampled_queries)
+                )
+                logger.info(
+                    "Remaining (Unsampled) Queries: %s",
+                    len(self.data_bucket.unsampled_queries),
+                )
+                (
+                    eval_stats,
+                    preds_single,
+                    preds_multi,
+                    domain_indices,
+                ) = self.mindmeld_classifier.train(
+                    data_bucket=self.data_bucket,
+                    return_preds_multi=STRATEGY_ABRIDGED[strategy]
+                    in ["ds", "ens", "kld"],
+                )
+                if not selection_mode:
+                    self.output_manager.update_accuracies_json(
+                        strategy, epoch, iteration, eval_stats
+                    )
+                    if self.save_sampled_queries:
+                        self.output_manager.update_selected_queries_json(
+                            strategy,
+                            epoch,
+                            iteration,
+                            self.data_bucket.newly_sampled_queries,
+                        )
+                    if self.early_stopping_window > 0:
+                        early_stop = self.output_manager.check_early_stopping(
+                            strategy, self.early_stopping_window
+                        )
+                num_unsampled = len(self.data_bucket.unsampled_queries)
+                if num_unsampled > 0:
+                    sampling_size = (
+                        self.batch_size
+                        if num_unsampled > self.batch_size
+                        else num_unsampled
+                    )
+                    heuristic = HeuristicsFactory.get_heuristic(strategy, sampling_size)
+                    (
+                        self.data_bucket.newly_sampled_queries,
+                        self.data_bucket.sampled_queries,
+                        self.data_bucket.unsampled_queries,
+                        _,
+                    ) = heuristic.sample(
+                        sampled=self.data_bucket.sampled_queries,
+                        unsampled=self.data_bucket.unsampled_queries,
+                        preds_single=preds_single,
+                        preds_multi=preds_multi,
+                        domain_indices=domain_indices,
+                    )
+                if selection_mode or early_stop:
+                    return
+                iteration += 1
```

Comment:
I would like more descriptions in this method. Its not entirely clear what this function is doing apart from the train step since it's doing more ops than that.